### ISSUE TITLE
Compile OCaml with Melange

### ISSUE THREAD

**Issue by EduardoRFS** [melange-re/melange#164]
I'm opening this issue because hopefully we can compile OCaml with Melange soon. And this is a good signal that we support OCaml.

## Known issues

- [ ] Nativeint is not supported
- [ ] `bsc file.ml` ignores `file.mli` leading to weak types
- [ ] `symtable.ml` uses `Obj.new_block` and `Obj.set_field`

**Comment by EduardoRFS** [comment]
After a lot of hacks, ALMOST

```shell
$ /usr/bin/node driver/main.js 
/home/eduardo/reason/ocaml/_esy/ocaml-development/store/b/melange-ed5e2880/default/lib/js/caml_external_polyfill.js:16
    throw new Error(s + " not polyfilled by ReScript yet\n")
    ^

Error: caml_weak_create not polyfilled by ReScript yet
```

**Comment by TheSpyder** [comment]
Some of this will require dropping IE11 support, but that's probably ok.

>  Weak is not supported

This can likely be done with a combination of [WeakMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap), [WeakSet](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet) and [WeakRef](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef).

Note that the MDN WeakRef docs states Safari is not supported but it does work (I've filed a bug).

>  Nativeint is not supported

There was previously a nativeint implementation but it wasn't correct. It should be possible to implement it using [Bigint](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt), but instead of attempting this it was removed.
https://github.com/rescript-lang/rescript-compiler/issues/4440

Looking through other ReScript issues that mention nativeint, one interesting complication is likely to be pattern matching:

```ocaml
let f = function 
  | 1n
  | 2n -> 3n
  | x -> Nativeint.add 3n x
```

So this may involve more than just implementing the basic nativeint functions.

**Comment by EduardoRFS** [comment]
@TheSpyder I don't think we need to use `BigInt` for Nativeint as our Nativeint is 32 bits and that can be encoded in OCaml easily, my guess is that it can behave exactly like `Int`.

And it will behave exactly like OCaml 32bits platforms, my guess is that we can even generate binaries for it. It would be also cool to have support to `Nativeint` as 64bits but that is a long shot. But also it's one my dream to make the OCaml backend able to compile from 32bits -> 64bits and vice versa, so the Nativeint as 64bits may be not needed.

`WeakMap` seems to be supported by IE11 so we can probably use it and it seems enough to encode the module `Weak` in OCaml

**Comment by TheSpyder** [comment]
JS isn’t a 32 bit platform though, if there’s no advantage over `Int` why would anyone want to use it. Bigint gives us 64 bit support.

**Comment by EduardoRFS** [comment]
@TheSpyder

> JS isn’t a 32 bit platform though, if there’s no advantage over Int why would anyone want to use it. 

Because `Nativeint` is portable, that's the same reason why people would use `Nativeint` in native OCaml. `Nativeint` will always describe the natural support for integers on a platform.

And JS is a 32bit platform, the only kind of integers supported naturally by JS is 32bits.

**Comment by TheSpyder** [comment]
> JS is a 32bit platform, the only kind of integers supported naturally by JS is 32bits.

That’s not true with bigint, it was added to the [primitive data types](https://developer.mozilla.org/en-US/docs/Glossary/Primitive), and arguably it’s not true at all - JS integers are [64 bit](https://developer.mozilla.org/en-US/docs/Glossary/Number) and the maximum value is an obscene [2^1024](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_VALUE) (although the largest "safe integer" is [2^53 - 1](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER), anything bigger is an approximation).

I think it's safe to say it is supported "naturally", whatever that means, even if some older devices don't implement bigint (a self-solving problem).

**Comment by EduardoRFS** [comment]
@TheSpyder BigInt is not a type of integer in the low level sense of it.

> JS integers are 64 bit

JS integers are not 64bit, `number` is a float of 64bits like in OCaml. That's why 2^53-1 is the safe integer, because that's how much a float can safely encode.

Remember `typeof` in JS doesn't give you the type of something in memory, it only gives a semantic type, for example a function is actually described as an object per ECMA262 with a [[Call]] symbol. A number is a 64bits float, but because of the bitwise OP spec, which describes operations on numbers by first casting them to 32bits integers, basically every JS engine can treat numbers as 32bits integers.

https://tc39.es/ecma262/#sec-numberbitwiseop

So the natural range of numbers which JS can process the fastest, is only 32bits integers, and I believe that as soon as you cross the 32bits range for integers you get a performance penalty, that's also why the `Int` module in JS is 32 and in OCaml 31 or 63.

**Comment by TheSpyder** [comment]
You're right, I forgot the math operations on numbers truncate to 32 bit (I hit this in my project at work). I realise `n | 0` is used all over generated code to truncate to 32 bits. But bigint supports, as far as I can tell, every number operator. Including bitwise without truncation.
https://tc39.es/ecma262/#sec-bigintbitwiseop

I certainly haven't tested whether bigint is good enough to encode `Nativeint` at 64 bit. But my expectation is that it can.

**Comment by TheSpyder** [comment]
also, having `500n` in OCaml compile to `500n` in JS instead of `500 | 0` will add to the "cool readable output" factor.

**Comment by EduardoRFS** [comment]
@TheSpyder `n | 0` is also a performance feature. And it actually works, engines can easily optimize it, is an old trick from asm.js time.

Encoding `Nativeint` as 64bit wouldn't be helpful to achieve codegen,

The problem of bigint is the platform requirement, it cannot be easily transpiled so adding a dependency on it is really bad. But also the syntax conflict is truly a shame, `500n` in Reason means something completely different ;/
